kolibri/core/assets/src/api_resource.js

```diff
@@ -39,24 +42,87 @@ class Model {
    * returns, otherwise reject is called with the response object.
    */
   fetch(params = {}, force = false) {
-    if (!force && this.synced) {
-      return Promise.resolve(this.attributes);
-    }
-    this.synced = false;
-    return new Promise((resolve, reject) => {
-      // Do a fetch on the URL.
-      client({ path: this.url, params }).then((response) => {
-        // Set the retrieved Object onto the Model instance.
-        this.set(response.entity);
-        // Flag that the Model has been fetched.
-        this.synced = true;
-        // Resolve the promise with the attributes of the Model.
-        resolve(this.attributes);
-      }, (response) => {
-        logging.error('An error occurred', response);
-        reject(response);
+    const promise = new Promise((resolve, reject) => {
+      Promise.all(this.promises).then(() => {
+        if (!force && this.synced) {
+          resolve(this.attributes);
+        } else {
+          this.synced = false;
+          // Do a fetch on the URL.
+          client({ path: this.url, params }).then((response) => {
+            // Set the retrieved Object onto the Model instance.
+            this.set(response.entity);
+            // Flag that the Model has been fetched.
+            this.synced = true;
+            // Resolve the promise with the attributes of the Model.
+            resolve(this.attributes);
+            // Clean up the reference to this promise
+            this.promises.splice(this.promises.indexOf(promise), 1);
+          }, (response) => {
+            logging.error('An error occurred', response);
+            reject(response);
+            // Clean up the reference to this promise
+            this.promises.splice(this.promises.indexOf(promise), 1);
+          });
+        }
       });
     });
+    this.promises.push(promise);
+    return promise;
+  }
+
+  /**
+   * Method to save data to the server for this particular model.
+   * @param {object} attrs - an object of attributes to be saved on the model.
+   * @returns {Promise} - Promise is resolved with Model attributes when the XHR successfully
+   * returns, otherwise reject is called with the response object.
+   */
+  save(attrs) {
+    const promise = new Promise((resolve, reject) => {
+      Promise.all(this.promises).then(() => {
+        let payload = {};
+        if (this.synced) {
+          // Model is synced with the server, so we can do dirty checking.
+          Object.keys(attrs).forEach((key) => {
+            if (attrs[key] !== this.attributes[key]) {
```

Comment:
Quick skim didn't show what would happen if the resulting payload is empty (i.e. nothing was dirty). Presumably could avoid firing a request at all, in that case, and queued promise would be a nullop. Having that in place could mean we could be a bit more liberal in firing saves, since they'd only actually fire off requests if needed.